### PR TITLE
[DellEMC] S6100 Last Reboot Reason Thermal Support

### PR DESCRIPTION
**Pull request by santhosh-kt**
**- What I did**

- Added support for Thermal event in Last Reboot Reason "show reboot-cause" command.
  Updated:
- Handled dependency based changes(Between process-reboot-cause and s6100_platform.sh).
- Handled newer bios based changes(holding previous reset register values and retrieve them in case of secondary bios boot).

**- How I did it**
On branch sonic-s6100-lrr
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

        modified:   files/image_config/platform/rc.local
        new file:   platform/broadcom/sonic-platform-modules-dell/common/nvram_rd_wr.py
        modified:   platform/broadcom/sonic-platform-modules-dell/debian/platform-modules-s6100.install
        new file:   platform/broadcom/sonic-platform-modules-dell/s6100/scripts/fast-reboot_plugin
        modified:   platform/broadcom/sonic-platform-modules-dell/s6100/scripts/platform_reboot_override
        modified:   platform/broadcom/sonic-platform-modules-dell/s6100/scripts/s6100_platform.sh
        new file:   platform/broadcom/sonic-platform-modules-dell/s6100/scripts/track_reboot_reason.sh
        new file:   platform/broadcom/sonic-platform-modules-dell/s6100/scripts/warm-reboot_plugin
        modified:   platform/broadcom/sonic-platform-modules-dell/s6100/sonic_platform/chassis.py
        modified:   platform/broadcom/sonic-platform-modules-dell/s6100/sonic_platform/eeprom.py
        new file:   platform/broadcom/sonic-platform-modules-dell/s6100/systemd/s6100-reboot-cause.service

**- How to verify it**
Added the Unit Test Results.

[All Test case result.txt](https://github.com/Azure/sonic-buildimage/files/4041135/All.Test.case.result.txt) - Updated
[Secondary bios boot - test results.txt](https://github.com/Azure/sonic-buildimage/files/4041136/secondary.bios.boot.-.test.results.txt)
[Unit test - Last Reboot Reason - S6100.txt](https://github.com/Azure/sonic-buildimage/files/4041137/Unit.test.-.Last.Reboot.Reason.-.S6100.txt) - Updated
[environment.txt](https://github.com/Azure/sonic-buildimage/files/4041138/environment.txt)

**- Description for the changelog**

[DellEMC] S6100 Last Reboot Reason Thermal Support and Dependency based changes

Depends on [Azure/sonic-utilities#740](https://github.com/Azure/sonic-utilities/pull/740)
**DO NOT MERGE UNTIL ABOVE PR IS MERGED**
Update: Azure/sonic-utilities#740 is merged.

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by santhosh-kt** [comment]
Azure/sonic-utilities#740 PR is merged. Please review the above changes.

**Comment by sujinmkang** [comment]
retest this please

**Comment by santhosh-kt** [comment]
Internal discussion is being taken place, please hold the review till further changes applied to this PR.

**Comment by santhosh-kt** [comment]
retest vsimage please

**Comment by santhosh-kt** [comment]
retest vsimage please

**Comment by santhosh-kt** [comment]
Added all the changes. Please continue to review the changes.

**Comment by santhosh-kt** [comment]
retest this please

**Comment by santhosh-kt** [comment]
retest broadcom please

**Comment by sujinmkang** [comment]
retest this please

**Comment by sujinmkang** [comment]
retest vsimage please

File: platform/broadcom/sonic-platform-modules-dell/s6100/sonic_platform/eeprom.py [hunk 1]
```diff
@@ -35,9 +35,13 @@ def __init__(self, i2c_line=0, iom_eeprom=False):
...
         except:
-            self.eeprom_data = "N/A"
             if not self.is_module:
-                raise RuntimeError("Eeprom is not Programmed")
+                try:
+                    self.eeprom_data = self.read_eeprom()
+                except:
+                    raise RuntimeError("Eeprom is not Programmed")
+            else:
+                self.eeprom_data = "N/A"
```
[the remark this answers is not in the document]
Comment (santhosh-kt) [reply platform/broadcom/sonic-platform-modules-dell/s6100/sonic_platform/eeprom.py]: Due to timing issues, we observed EEPROM read failures during the boot time and we handled that erroneous case here.

File: platform/broadcom/sonic-platform-modules-dell/s6100/scripts/track_reboot_reason.sh [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    fi
+
+    echo 0
+    return
+}
+
+_is_unknown_reset(){
+    if [[ -f $REBOOT_CAUSE_FILE ]]; then
+	if [[ $1 = 0 ]]; then
+	    echo "Unknown software reboot" > $REBOOT_CAUSE_FILE
```
[the remark this answers is not in the document]
Comment (santhosh-kt) [reply platform/broadcom/sonic-platform-modules-dell/s6100/scripts/track_reboot_reason.sh]: Usually all software reboot cases are handled by SONiC itself, but other cases like kernel panic, etc wont be handled in SONiC. At present, for all software resets other than warm-reboot, fast-reboot and reboot, the value "Unknown software reboot" would be set in reboot-cause.txt file.

**Comment by santhosh-kt** [comment]
It looks like, Azure/sonic-utilities#740 PR is not synced to master branch. s6100-reboot-cause.service requires this PR, that will handle warm-reboot, fast-reboot and reboot cases in 2.0 API.

Attached logs regarding the plugin - [warm-reboot.txt](https://github.com/Azure/sonic-buildimage/files/4116873/warm-reboot.txt)

This Azure/sonic-utilities#740 PR is already picked in 201911 but not in master yet.

**Comment by jleveque** [comment]
@santhosh-kt: FYI, https://github.com/Azure/sonic-utilities/pull/740 made it into the master branch today with this sonic-utilities submodule update: https://github.com/Azure/sonic-buildimage/pull/4080

File: files/image_config/platform/rc.local [hunk 3]
```diff
@@ -240,6 +240,13 @@ if [ -f $FIRST_BOOT_FILE ]; then
     # Notify firstboot to Platform, to use it for reboot-cause
     touch /tmp/notify_firstboot_to_platform
 
+    # create reboot-cause/platform/$platform/ directory
```
Comment (jleveque) [inline files/image_config/platform/rc.local]: The path in this comment doesn't align with the actual directory you're creating below. I think you should also note that this directory can be used as an intermediate directory by platform vendors.

File: platform/broadcom/sonic-platform-modules-dell/s6100/sonic_platform/eeprom.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jleveque) [reply platform/broadcom/sonic-platform-modules-dell/s6100/sonic_platform/eeprom.py]: Then should this be submitted as a separate PR if it is not related to reboot cause?

[the remark this answers is not in the document]
Comment (santhosh-kt) [reply platform/broadcom/sonic-platform-modules-dell/s6100/sonic_platform/eeprom.py]: It is indirectly related to reboot cause as process-reboot-cause gets failed when trying to initialize Chassis() where EEPROM read failure happened rarely due to timing issues. 
If you still prefer to have separate PR, we can raise it.

File: files/image_config/platform/rc.local (hunk 3 above)
Comment (santhosh-kt) [reply files/image_config/platform/rc.local]: Will update the correct path in the comment.

File: files/image_config/platform/rc.local [hunk 4]
```diff
@@ -240,6 +240,13 @@ if [ -f $FIRST_BOOT_FILE ]; then
     # Notify firstboot to Platform, to use it for reboot-cause
     touch /tmp/notify_firstboot_to_platform
 
+    # create /host/reboot-cause/platform/ directory
+    # /host/reboot-cause/platform/reboot-reason
+    # will be used to track last reboot reason
```
Comment (jleveque) [inline files/image_config/platform/rc.local]: This new 'platform' directory is not used by all platforms. I think this line should reflect that. Something like,

`# can be used to track last reboot reason by some platforms`

Comment (santhosh-kt) [reply files/image_config/platform/rc.local]: Updated

File: platform/broadcom/sonic-platform-modules-dell/s6100/sonic_platform/eeprom.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jleveque) [reply platform/broadcom/sonic-platform-modules-dell/s6100/sonic_platform/eeprom.py]: If it is indirectly related to reboot cause, it can stay.

**Comment by jleveque** [comment]
Retest vsimage please

**Comment by sujinmkang** [comment]
test vsimage please

**Comment by sujinmkang** [comment]
Retest vsimage please

File: files/image_config/platform/rc.local [hunk 5]
```diff
@@ -240,6 +240,12 @@ if [ -f $FIRST_BOOT_FILE ]; then
     # Notify firstboot to Platform, to use it for reboot-cause
     touch /tmp/notify_firstboot_to_platform
 
+    # Create /host/reboot-cause/platform/ directory
+    # can be used to track last reboot reason by some platforms
+    if [ ! -d /host/reboot-cause/platform ]; then
+	mkdir -p /host/reboot-cause/platform
```
Comment (jleveque) [inline files/image_config/platform/rc.local]: Please replace tabs with 4 spaces

Comment (santhosh-kt) [reply files/image_config/platform/rc.local]: Updated.

**Comment by sujinmkang** [comment]
test vsimage please

**Comment by sujinmkang** [comment]
retest vsimage please

**Comment by santhosh-kt** [comment]
These changes are not ported to 201911 yet. Needed to be ported.

**Pull request merged**